### PR TITLE
Support custom wrappers

### PR DESCRIPTION
**Pull request by selsamman**
Add support for libraries that require custom wrappers for components and a custom useTransition.  Proxy based libraries often need a wrapper like observer (mobx, proxily) and proxily, which requires a custom useTransition.  They are passed as parameters to common with defaults so no changes are required for other libraries. Added proxily which uses both new parameters.  Discussed in #47

### CODE REVIEW

**Comment by dai-shi** [comment]
Would you merge the main branch?
(Also, we would probably redesign the entire tests to match with the latest react 18 features, and everything can be deleted then.)

File: src/common.js
Comment (dai-shi) [inline src/common.js:86]: custom wrapper might be be acceptable, but custom useTransition doesn't feel right. 🤔

**Comment by dai-shi** [comment]
As I implied, we will probably redesign tests without useTransition. So, let me hold this PR or close it. (Sorry, I should have notice it before asking the merge.)

**Comment by selsamman** [comment]
Sorry is useTransition going away?  I could not find anything on that in the R18 WG discussion.

As for why the need for a custom useTranstion it is that in order to create snapshots for the transition one has to know that the code executing is part of the transition.  The two options I could think of to detect a transition were:

```
startTransition( () => startObservableTransition( () => { // stuff in transition } ))
```

or

```
startObservableTransition (() => { // stuff in transition })  // does both
```

Everything related to R18 is still experimental and I am not wedded to either but the 2nd felt cleaner.   The 1st would not require this change to common.  In any event if you are planning on updating the test to not incorporate useTransition in the project anymore then no point merging this PR.  I would imagine you will include a custom component wrapper when you rewrite it.

**Comment by dai-shi** [comment]
> Sorry is useTransition going away? I could not find anything on that in the R18 WG discussion.

Sorry, I'm still learning and can't explain well. Not going away, but less important with uSES? https://github.com/reactwg/react-18/discussions/86

> startTransition( () => startObservableTransition( () => { // stuff in transition } ))

This seems good to me. I did it in `use-context-selector`.

> I would imagine you will include a custom component wrapper when you rewrite it.

I can do it now. Or, if you shrink this PR to add `memo = React.memo` in `createApp` without adding proxily, I can merge.

**Comment by selsamman** [comment]
Won't support for transitions remain important in determining whether a library works with concurrent rendering? State branching is part of concurrent rendering.  Can it be adequately tested without startTransition?

**Comment by dai-shi** [comment]
yeah, I'm rather interested in testing state branching. (as tearing would be just solved with uSES.)

**Comment by selsamman** [comment]
Can't seem to ditch restoring the ^M in the README.   Hopefully that is OK.

I do think you have hit the sweet spot with the setup you have now on testing state branching.  

And yes I hope uSES will resolve tearing though I worry about forcing synchronous renders causing us to loss on some  use cases like an input field keeping up with typing when the updates cause expensive renders.

**Review by dai-shi — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**
